### DISCUSSION TITLE
I want to get the reflectance of each wavelength on a material surface.

### DISCUSSION THREAD

**Discussion by Bambootree0818** [mitsuba-renderer/mitsuba3#1278] [Q&A]
Hello!

I am currently looking to get the reflectance of a material (such as Principled BRDF) in 10nm intervals from 400nm to 700nm.

The image is like using a spectrophotometer to obtain the spectral spectrum of an object's surface.

Is that possible with mitsuba3?
I looked for an API to do that in the references but could not find it successfully as I am a newbie to mitsuba3.

thank you.

**Comment by njroussel** [comment]
Hi @Bambootree0818 

Yes, this is possible. 
See this [section of a tutorial](https://mitsuba.readthedocs.io/en/stable/src/others/bsdf_deep_dive.html#Vectorized-evaluation-of-the-BSDF) on how to evaluate a BSDF for a set of incident & outgoing directions. You'll need to make one small change, which is to specify the wavelengths that you want to evaluate for -- this is not needed in the tutorial because it's using and `rgb` variant. This can be done by setting [`si.wavelgenths`](https://mitsuba.readthedocs.io/en/stable/src/api_reference.html#mitsuba.Interaction3f.wavelengths). Note that a single ray carries 4 wavelgenths in `_spectral` variants, so evaluation of `BSDF.eval()` will also process 4 wavelengths per set of directions.

**Reply by Bambootree0818** [reply]
Thank you very much for your reply. @njroussel 

I implemented the following code based on your suggestions. However, I did not get the output I wanted.

What I want to do is to use a spectrophotometer to measure the reflectance of an actual red paint sample for each wavelength, and compare the results with the simulation results on the mitsuba3.

Current problems are:

1. All the results are almost the same for each wavelength I used to evaluate the reflectance.

2. The reflectance range is not returned as 0-100, as in the measurement results (Header:Reflectance of each wavelength of a red paint sample using a spectrophotometer) of the actual sample. I'm hoping for an output like the image.

I would like to know if there is anything I'm misunderstanding or if there is something I'm doing wrong in this code.


### Code
```python
import mitsuba as mi
import numpy as np
import drjit as dr

mi.set_variant('scalar_spectral')

bsdf = mi.load_dict({
    'type': 'principled',
    'base_color': {
            'type': 'rgb',
            'value': [0.3391, 0.0008, 0.0195]
    },
    'metallic': 0.5,
    'specular': 0.5,
    'roughness': 0.5,
    'spec_tint': 0.0,
    'anisotropic': 0.0,
    'sheen': 0.0,
    'sheen_tint': 0.0,
    'clearcoat': 1.0,
    'clearcoat_gloss': 1.0,
    'spec_trans': 0.0
})

# Define a range of wavelengths
wavelengths = np.arange(400.0, 701.0, 20.0)  

def sph_to_dir(theta, phi):
    """Map spherical to Euclidean coordinates"""
    st, ct = dr.sincos(theta)
    sp, cp = dr.sincos(phi)
    return mi.Vector3f(cp * st, sp * st, ct)

# Calculate reflectance at each wavelength
reflectance_spectrum = []
for wl in wavelengths:
    # Create a (dummy) surface interaction to use for the evaluation of the BSDF
    si = dr.zeros(mi.SurfaceInteraction3f)

    # Specify an incident direction with 45 degrees elevation
    si.wi = sph_to_dir(dr.deg2rad(45.0), 0.0)

    #print(wl)
    #Define wavelength
    si.wavelengths = mi.scalar_spectral.Spectrum(wl)

    wo = sph_to_dir(dr.deg2rad(-45.0), 0.0)

    # Evaluate the whole array (18000 directions) at once
    value = bsdf.eval(mi.BSDFContext(), si, wo)
    #print(value)
    reflectance_spectrum.append(value)

# 結果を表示
for wl, refl in zip(wavelengths, reflectance_spectrum):
    print(f"Wavelength: {wl} nm, Reflectance: {refl}")
```

### Result
Wavelength: 400.0 nm, Reflectance: [146.8250732421875, 146.8250732421875, 146.8250732421875, 146.8250732421875]
Wavelength: 410.0 nm, Reflectance: [146.8179473876953, 146.8179473876953, 146.8179473876953, 146.8179473876953]
Wavelength: 420.0 nm, Reflectance: [146.81358337402344, 146.81358337402344, 146.81358337402344, 146.81358337402344]
Wavelength: 430.0 nm, Reflectance: [146.810791015625, 146.810791015625, 146.810791015625, 146.810791015625]
Wavelength: 440.0 nm, Reflectance: [146.80894470214844, 146.80894470214844, 146.80894470214844, 146.80894470214844]
Wavelength: 450.0 nm, Reflectance: [146.8077392578125, 146.8077392578125, 146.8077392578125, 146.8077392578125]
Wavelength: 460.0 nm, Reflectance: [146.80699157714844, 146.80699157714844, 146.80699157714844, 146.80699157714844]
Wavelength: 470.0 nm, Reflectance: [146.8065948486328, 146.8065948486328, 146.8065948486328, 146.8065948486328]
Wavelength: 480.0 nm, Reflectance: [146.80650329589844, 146.80650329589844, 146.80650329589844, 146.80650329589844]
Wavelength: 490.0 nm, Reflectance: [146.80673217773438, 146.80673217773438, 146.80673217773438, 146.80673217773438]
Wavelength: 500.0 nm, Reflectance: [146.80726623535156, 146.80726623535156, 146.80726623535156, 146.80726623535156]
Wavelength: 510.0 nm, Reflectance: [146.80819702148438, 146.80819702148438, 146.80819702148438, 146.80819702148438]
Wavelength: 520.0 nm, Reflectance: [146.8096160888672, 146.8096160888672, 146.8096160888672, 146.8096160888672]
Wavelength: 530.0 nm, Reflectance: [146.8118133544922, 146.8118133544922, 146.8118133544922, 146.8118133544922]
Wavelength: 540.0 nm, Reflectance: [146.81515502929688, 146.81515502929688, 146.81515502929688, 146.81515502929688]
Wavelength: 550.0 nm, Reflectance: [146.8204345703125, 146.8204345703125, 146.8204345703125, 146.8204345703125]
Wavelength: 560.0 nm, Reflectance: [146.82940673828125, 146.82940673828125, 146.82940673828125, 146.82940673828125]
Wavelength: 570.0 nm, Reflectance: [146.84603881835938, 146.84603881835938, 146.84603881835938, 146.84603881835938]
Wavelength: 580.0 nm, Reflectance: [146.8817138671875, 146.8817138671875, 146.8817138671875, 146.8817138671875]
Wavelength: 590.0 nm, Reflectance: [146.97291564941406, 146.97291564941406, 146.97291564941406, 146.97291564941406]
Wavelength: 600.0 nm, Reflectance: [147.2144012451172, 147.2144012451172, 147.2144012451172, 147.2144012451172]
Wavelength: 610.0 nm, Reflectance: [147.53622436523438, 147.53622436523438, 147.53622436523438, 147.53622436523438]
Wavelength: 620.0 nm, Reflectance: [147.68536376953125, 147.68536376953125, 147.68536376953125, 147.68536376953125]
Wavelength: 630.0 nm, Reflectance: [147.7357177734375, 147.7357177734375, 147.7357177734375, 147.7357177734375]
Wavelength: 640.0 nm, Reflectance: [147.7554473876953, 147.7554473876953, 147.7554473876953, 147.7554473876953]
Wavelength: 650.0 nm, Reflectance: [147.76463317871094, 147.76463317871094, 147.76463317871094, 147.76463317871094]
Wavelength: 660.0 nm, Reflectance: [147.76950073242188, 147.76950073242188, 147.76950073242188, 147.76950073242188]
Wavelength: 670.0 nm, Reflectance: [147.77232360839844, 147.77232360839844, 147.77232360839844, 147.77232360839844]
Wavelength: 680.0 nm, Reflectance: [147.77407836914062, 147.77407836914062, 147.77407836914062, 147.77407836914062]
Wavelength: 690.0 nm, Reflectance: [147.7752227783203, 147.7752227783203, 147.7752227783203, 147.7752227783203]
Wavelength: 700.0 nm, Reflectance: [147.7760009765625, 147.7760009765625, 147.7760009765625, 147.7760009765625]

### Reflectance of each wavelength of a red paint sample using a spectrophotometer
![スクリーンショット 2024-08-19 205326](https://github.com/user-attachments/assets/745cfeda-d0b3-464b-9e2c-7387fbb8c137)

**Reply by njroussel** [reply]
Do you want to measure the transmission?
You're currently setting the outgoing direction below the surface with `wo = sph_to_dir(dr.deg2rad(-45.0), 0.0)`, however you've also set the transmissive lobe to be zero with ` 'spec_trans': 0.0`. Fixing that that should give you reasonable values at least.

(I'm not sure why it isn't returning 0 currently though, we'll have a look...)

**Reply by Bambootree0818** [reply]
Thanks for your quick reply. @njroussel 
Sorry. I was very wrong.

The spectrophotometer used to measure the actual sample had an incident direction of 0° and a receiving angle of 45°, but I forgot to reflect that in the code. I fixed it.

The results are shown in the output section.
And I have one question here.
For each wavelength, 4 values are stored in the list. Does this mean that “BSDF.eval() evaluation also processes 4 wavelengths for one set of directions” as you indicated?

Ideally, I would like to get one reflectance per wavelength, like the spectrophotometer results in the image, so I would like to know how to do that.

### Code
```py
import mitsuba as mi
import numpy as np
import drjit as dr

mi.set_variant('scalar_spectral')

bsdf = mi.load_dict({
    'type': 'principled',
    'base_color': {
            'type': 'rgb',
            'value': [0.3391, 0.0008, 0.0195]
    },
    'metallic': 0.5,
    'specular': 0.5,
    'roughness': 0.5,
    'spec_tint': 0.0,
    'anisotropic': 0.0,
    'sheen': 0.0,
    'sheen_tint': 0.0,
    'clearcoat': 1.0,
    'clearcoat_gloss': 1.0,
    'spec_trans': 0.0
})

# Define a range of wavelengths
wavelengths = np.arange(400.0, 701.0, 20.0)  

def sph_to_dir(theta, phi):
    """Map spherical to Euclidean coordinates"""
    st, ct = dr.sincos(theta)
    sp, cp = dr.sincos(phi)
    return mi.Vector3f(cp * st, sp * st, ct)

# Calculate reflectance at each wavelength
reflectance_spectrum = []
for wl in wavelengths:
    # Create a (dummy) surface interaction to use for the evaluation of the BSDF
    si = dr.zeros(mi.SurfaceInteraction3f)

    # Specify an incident direction with 45 degrees elevation
    si.wi = sph_to_dir(0, 0)

    #print(wl)
    #Define wavelength
    si.wavelengths = mi.scalar_spectral.Spectrum(wl)

    wo = sph_to_dir(dr.deg2rad(45.0), 0.0)

    # Evaluate the whole array (18000 directions) at once
    value = bsdf.eval(mi.BSDFContext(), si, wo)
    #print(value)
    reflectance_spectrum.append(value)

# Result 
for wl, refl in zip(wavelengths, reflectance_spectrum):
    print(f"Wavelength: {wl} nm, Reflectance: {refl}")
```

### Result
Wavelength: 400.0 nm, Reflectance: [0.014543430879712105, 0.014543430879712105, 0.014543430879712105, 0.014543430879712105]
Wavelength: 420.0 nm, Reflectance: [0.009734602645039558, 0.009734602645039558, 0.009734602645039558, 0.009734602645039558]
Wavelength: 440.0 nm, Reflectance: [0.007863267324864864, 0.007863267324864864, 0.007863267324864864, 0.007863267324864864]
Wavelength: 460.0 nm, Reflectance: [0.007008153945207596, 0.007008153945207596, 0.007008153945207596, 0.007008153945207596]
Wavelength: 480.0 nm, Reflectance: [0.006625221576541662, 0.006625221576541662, 0.006625221576541662, 0.006625221576541662]
Wavelength: 500.0 nm, Reflectance: [0.006534837186336517, 0.006534837186336517, 0.006534837186336517, 0.006534837186336517]
Wavelength: 520.0 nm, Reflectance: [0.006700165569782257, 0.006700165569782257, 0.006700165569782257, 0.006700165569782257]
Wavelength: 540.0 nm, Reflectance: [0.007190178148448467, 0.007190178148448467, 0.007190178148448467, 0.007190178148448467]
Wavelength: 560.0 nm, Reflectance: [0.008251409977674484, 0.008251409977674484, 0.008251409977674484, 0.008251409977674484]
Wavelength: 580.0 nm, Reflectance: [0.010643448680639267, 0.010643448680639267, 0.010643448680639267, 0.010643448680639267]
Wavelength: 600.0 nm, Reflectance: [0.017279062420129776, 0.017279062420129776, 0.017279062420129776, 0.017279062420129776]
Wavelength: 620.0 nm, Reflectance: [0.04387344419956207, 0.04387344419956207, 0.04387344419956207, 0.04387344419956207]
Wavelength: 640.0 nm, Reflectance: [0.12665358185768127, 0.12665358185768127, 0.12665358185768127, 0.12665358185768127]
Wavelength: 660.0 nm, Reflectance: [0.16568569839000702, 0.16568569839000702, 0.16568569839000702, 0.16568569839000702]
Wavelength: 680.0 nm, Reflectance: [0.17343810200691223, 0.17343810200691223, 0.17343810200691223, 0.17343810200691223]
Wavelength: 700.0 nm, Reflectance: [0.1756662130355835, 0.1756662130355835, 0.1756662130355835, 0.1756662130355835]

### Reflectance of each wavelength of a red paint sample using a spectrophotometer
![スクリーンショット 2024-08-19 205326](https://github.com/user-attachments/assets/745cfeda-d0b3-464b-9e2c-7387fbb8c137)

**Reply by njroussel** [reply]
> Does this mean that “BSDF.eval() evaluation also processes 4 wavelengths for one set of directions” as you indicated?

Yes.

You can either index in to your output `bsdf.eval(...)[0]` and discard the 3 other wavelenghts, since they are indentical. Or you can split you range of wavelengths into chuncks of 4, and pass them 4 at a time like: 
```python
N_wavelengths = 20 # multiple of 4
wavelengths = np.arange(400.0, 701.0, N_wavelenghts)  
for i in range(N_wavelenghts)[::4]:
     (...)
     si.wavelengths = [wavelengths[i + 0], 
                       wavelengths[i + 1], 
                       wavelengths[i + 2], 
                       wavelengths[i + 3]]
     value = bsdf.eval(mi.BSDFContext(), si, wo)
```

**Reply by Bambootree0818** [reply]
I did not know that I could discard the three wavelengths I obtained. I have made a correction based on the code you suggested. It appears to be working well.

**I really appreciate your detailed explanations!** @njroussel 

However, another question has arisen.
I saw an article about using a BSDF measurement device to measure the BSDF of a real sample and fitting mitsuba3's material to it.
In that case, the function used to evaluate the BSDF of the material prepared by mitsuba3 was also bsdf.eval().

**Are the values output in my code this time also BSDF values?**

What I want is the reflectance of the material (**ratio of reflected light intensity to incident light intensity, range 0~1**), which seems different from the BSDF value (0 ~ ∞).

### Code
```py
import mitsuba as mi
import numpy as np
import drjit as dr

mi.set_variant('scalar_spectral')

bsdf = mi.load_dict({
    'type': 'principled',
    'base_color': {
            'type': 'rgb',
            'value': [1.0, 0.0008, 0.0095]
    },
    'metallic': 0.3,
    'specular': 0.3,
    'roughness': 0.3,
    'spec_tint': 0.0,
    'anisotropic': 0.0,
    'sheen': 0.0,
    'sheen_tint': 0.0,
    'clearcoat': 0.0,
    'clearcoat_gloss': 0.0,
    'spec_trans': 0.0
})

# Define a range of wavelengths
N_wavelengths = 20
wavelengths = np.arange(400.0, 701.0, N_wavelengths)  

#print(wavelengths)

def sph_to_dir(theta, phi):
    """Map spherical to Euclidean coordinates"""
    st, ct = dr.sincos(theta)
    sp, cp = dr.sincos(phi)
    return mi.Vector3f(cp * st, sp * st, ct)

# Calculate reflectance at each wavelength
reflectance_spectrum = []
for i in range(0, len(wavelengths) - 3, 4):
    # Create a (dummy) surface interaction to use for the evaluation of the BSDF
    si = dr.zeros(mi.SurfaceInteraction3f)

    # Specify an incident direction with 45 degrees elevation
    si.wi = sph_to_dir(0, 0)

    #Define wavelength
    si.wavelengths = [wavelengths[i],
                      wavelengths[i + 1],
                      wavelengths[i + 2],
                      wavelengths[i + 3]]
    
    #print(si.wavelengths)

    wo = sph_to_dir(dr.deg2rad(45.0), 0.0)

    # Evaluate the whole array (18000 directions) at once
    value = bsdf.eval(mi.BSDFContext(), si, wo)
    #print(value)
    reflectance_spectrum.append(value)

# Result
for i in range(0, len(wavelengths), 4):
    wl_slice = wavelengths[i:i+4]  
    refl = reflectance_spectrum[i // 4]  

    print(f"Wavelengths: {wl_slice} nm, Reflectance: {refl}")

```

### Result
Wavelengths: [400. 420. 440. 460.] nm, Reflectance: [0.12188934534788132, 0.011093857698142529, 0.0035530731547623873, 0.002195152919739485]
Wavelengths: [480. 500. 520. 540.] nm, Reflectance: [0.001778108300641179, 0.0016859248280525208, 0.0018263242673128843, 0.0023457196075469255]
Wavelengths: [560. 580. 600. 620.] nm, Reflectance: [0.004119507968425751, 0.016360051929950714, 0.14292317628860474, 0.1637125313282013]
Wavelengths: [640. 660. 680. 700.] nm, Reflectance: [0.16548742353916168, 0.16591322422027588, 0.16606703400611877, 0.16613605618476868]

### Reflectance of each wavelength of a red paint sample using a spectrophotometer
![スクリーンショット 2024-08-19 205326](https://github.com/user-attachments/assets/745cfeda-d0b3-464b-9e2c-7387fbb8c137)

**Reply by njroussel** [reply]
Hi again @Bambootree0818 

> Are the values output in my code this time also BSDF values?
> What I want is the reflectance of the material (ratio of reflected light intensity to incident light intensity, range 0~1), which seems different from the BSDF value (0 ~ ∞).

\~_Yes_\~, for an energy conserving BSDF that should be the case. Note that, as mentionned in [`eval`'s documentation](https://mitsuba.readthedocs.io/en/stable/src/api_reference.html#mitsuba.BSDF.eval), the output is already multiplied by the cosine foreshortening term to take into account the projected solid angle of your input directions.

**Reply by Bambootree0818** [reply]
Hello @njroussel . Sorry for the late reply.

Little by little, I learned more about optics and mitsuba3. Thanks for the detailed explanation.

Thanks to you, I was able to get the reflectance in a specific direction using bsdf.eval().
Sorry to keep repeating this, but there is one more thing I am wondering about.

I have set the material color to be strongly red, like [1.0, 0.0, 0.0], yet the long wavelength portion output by bsdf.eval() is small.
For example
> Wavelengths: [640. 660. 680. 700.] nm, Reflectance: [0.1654076874256134, 0.16587819159030914, 0.16604872047901154, 0.1661253422498703]

**Could you guess why this is?**
I had imagined that if the material was red, the long wavelength band would show large values like 0.7,0.8

I could confirm that bsdf.eval_diffuse_reflectance() can be used to obtain values close to the actual sample measurements, but I could not understand the difference in specifications between it and bsdf.eval(). And I'm not sure if this is even appropriate to reproduce the behavior of a real-world spectrophotometer.

### Code
```py
import mitsuba as mi
import numpy as np
import drjit as dr

mi.set_variant('scalar_spectral')

bsdf = mi.load_dict({
    'type': 'principled',
    'base_color': {
            'type': 'srgb',
            'color': [1.0,0.0039,0.0092]
    },
    'metallic': 0.3,
    'specular': 0.3,
    'roughness': 0.3,
    'spec_tint': 0.0,
    'anisotropic': 0.0,
    'sheen': 0.0,
    'sheen_tint': 0.0,
    'clearcoat': 0.0,
    'clearcoat_gloss': 0.0,
    'spec_trans': 0.0
})

# Define a range of wavelengths
N_wavelengths = 20
wavelengths = np.arange(400.0, 701.0, N_wavelengths)  

#print(wavelengths)

def sph_to_dir(theta, phi):
    """Map spherical to Euclidean coordinates"""
    st, ct = dr.sincos(theta)
    sp, cp = dr.sincos(phi)
    return mi.Vector3f(cp * st, sp * st, ct)

# Calculate reflectance at each wavelength
reflectance_spectrum = []
for i in range(0, len(wavelengths) - 3, 4):
    # Create a (dummy) surface interaction to use for the evaluation of the BSDF
    si = dr.zeros(mi.SurfaceInteraction3f)

    # Specify an incident direction with 0 degrees elevation
    si.wi = sph_to_dir(dr.deg2rad(0.0), 0)
    
    #Define wavelength
    si.wavelengths = [wavelengths[i],
                      wavelengths[i + 1],
                      wavelengths[i + 2],
                      wavelengths[i + 3]]
    
    #print(si.wavelengths)

    wo = sph_to_dir(dr.deg2rad(45.0), 0.0)

    value = bsdf.eval(mi.BSDFContext(),si,wo)
    #print(value)
    reflectance_spectrum.append(value)

# display result
for i in range(0, len(wavelengths), 4):
    wl_slice = wavelengths[i:i+4] 
    refl = reflectance_spectrum[i // 4] 

    print(f"Wavelengths: {wl_slice} nm, Reflectance: {refl}")

```

### Result
Wavelengths: [400. 420. 440. 460.] nm, Reflectance: [0.110493503510952, 0.010803439654409885, 0.003702295944094658, 0.00232458533719182]
Wavelengths: [480. 500. 520. 540.] nm, Reflectance: [0.0018935453845188022, 0.001802740152925253, 0.0019635080825537443, 0.0025414787232875824]
Wavelengths: [560. 580. 600. 620.] nm, Reflectance: [0.004503182135522366, 0.017773447558283806, 0.14171378314495087, 0.16346140205860138]
Wavelengths: [640. 660. 680. 700.] nm, Reflectance: [0.1654076874256134, 0.16587819159030914, 0.16604872047901154, 0.1661253422498703]

### Reflectance of each wavelength of a red paint sample using a spectrophotometer
![スクリーンショット 2024-08-19 205326](https://github.com/user-attachments/assets/745cfeda-d0b3-464b-9e2c-7387fbb8c137)

**Reply by njroussel** [reply]
Firstly, this can largely depend on the  spectral upsampling scheme used (RGB to spectral conversion). Mitsuba uses the following work: [A Low-Dimensional Function Space for Efficient Spectral Upsampling](https://rgl.epfl.ch/publications/Jakob2019Spectral).

Secondly, for the configuration you've sent, an RGB evaluation will return `[0.225085, 0.000877831, 0.00207078]` which lost most of its red component - this is due to the angles and properties of the material.
